pinot-core/src/main/java/org/apache/pinot/core/data/manager/BaseTableDataManager.java

```diff
@@ -1046,6 +1064,261 @@ public boolean needReloadSegments()
+        // refresh is needed, since table has it enabled.
+        boolean incompatible = colHasDictionary || DictionaryIndexType.ignoreDictionaryOverride(
+            tableConfig.getIndexingConfig().isOptimizeDictionary(),
+            tableConfig.getIndexingConfig().isOptimizeDictionaryForMetrics(),
+            tableConfig.getIndexingConfig().getNoDictionarySizeRatioThreshold(),
+            tableConfig.getIndexingConfig().getNoDictionaryCardinalityRatioThreshold(), fieldSpecInSchema,
+            indexConfigsMap.get(columnName), columnMetadata.getCardinality(), columnMetadata.getTotalNumberOfEntries());
+        if (incompatible) {
+          LOGGER.debug("tableNameWithType: {}, segmentName: {}, change: dictionary encoding,", tableNameWithType,
+              segmentName);
+          return new NeedRefreshResponse(segmentName, true, "dictionary encoding changed: " + columnName);
+        } else {
+          LOGGER.debug("tableNameWithType: {}, segmentName: {}, no change as dictionary overrides applied to col: {}",
+              tableNameWithType, segmentName, columnName);
+        }
+      }
+
+      // Sorted column not sorted
+      if (columnName.equals(sortedColumn) && !columnMetadata.isSorted()) {
+        LOGGER.debug("tableNameWithType: {}, segmentName: {}, change: sort column", tableNameWithType, segmentName);
+        return new NeedRefreshResponse(segmentName, true, "sort column changed: " + columnName);
+      }
+
+      if (Objects.isNull(source.getBloomFilter()) == bloomFilters.contains(columnName)) {
+        LOGGER.debug("tableNameWithType: {}, segmentName: {}, change: bloom filter changed", tableNameWithType,
+            segmentName);
+        return new NeedRefreshResponse(segmentName, true, "bloom filter changed: " + columnName);
+      }
+
+      if (Objects.isNull(source.getJsonIndex()) == jsonIndex.contains(columnName)) {
+        LOGGER.debug("tableNameWithType: {}, segmentName: {}, change: json index changed", tableNameWithType,
+            segmentName);
+        return new NeedRefreshResponse(segmentName, true, "json index changed: " + columnName);
+      }
+
+      if (Objects.isNull(source.getTextIndex()) == textIndexes.contains(columnName)) {
+        LOGGER.debug("tableNameWithType: {}, segmentName: {}, change: text index changed", tableNameWithType,
+            segmentName);
+        return new NeedRefreshResponse(segmentName, true, "text index changed: " + columnName);
+      }
+
+      if (Objects.isNull(source.getFSTIndex()) == fstIndexes.contains(columnName)) {
+        LOGGER.debug("tableNameWithType: {}, segmentName: {}, change: fst index changed", tableNameWithType,
+            segmentName);
+        return new NeedRefreshResponse(segmentName, true, "fst index changed: " + columnName);
+      }
+
+      if (Objects.isNull(source.getH3Index()) == h3Indexes.contains(columnName)) {
+        LOGGER.debug("tableNameWithType: {}, segmentName: {}, change: h3 index changed", tableNameWithType,
```

Comment:
nit: Can we improve this log to also have columnName for which the index was changed

Reply:
Added `columnName` to all segments.